core/src/main/clojure/xtdb/metadata.clj

```diff
@@ -311,23 +315,47 @@
     {:col-names (into #{} col-names)
      :page-idx-cache page-idx-cache}))
 
-(defn ->table-metadata ^xtdb.metadata.ITableMetadata [^IVectorReader metadata-reader, {:keys [col-names ^Map page-idx-cache]}]
-  (reify ITableMetadata
-    (metadataReader [_] metadata-reader)
-    (columnNames [_] col-names)
-    (rowIndex [_ col-name page-idx] (.get page-idx-cache [col-name page-idx]))
-    (iidBloomBitmap [_ page-idx]
-      (let [bloom-rdr (-> (.structKeyReader metadata-reader "columns")
-                          (.listElementReader)
-                          (.structKeyReader "bloom"))]
+(defrecord TableMetadata [^HashTrie trie
+                          ^RelationReader meta-rel-reader
```

Comment:
latter is reading the leaves, IIUC?

Reply:
I essentially keep `meta-rel-reader` around for closing and `metadata-rdr` for reading the leafs. Wanted to avoid grabbing the `leaf-reader` every time, but agree it's a bit redundant.